### DISCUSSION TITLE
Roadmap:: coding/Design opinions

### DISCUSSION THREAD

**Discussion by dcowden** [CadQuery/cadquery#1109] [Ideas]
Roadmap step 3: consensus on key coding design rules for this phase. 

Note: this is NOT the same as the overall project-- this is specificly the rules/design opinions we want to use during design for the roadmap items.

Proposed rules are in the roadmap document: https://github.com/CadQuery/cadquery/wiki/RoadMap#codingdesign-opinions

**Comment by jmwright** [comment]
The [discussion on the rules of engagement](https://github.com/CadQuery/cadquery/discussions/1107#discussioncomment-3026296) suggested that items on converting to PEP8/P484 be put here. Below is an older issue related to PEP compatibilities.

https://github.com/CadQuery/cadquery/issues/195#issuecomment-628612589

**Comment by jmwright** [comment]
> 4. Lets try to move to immutable objects and eliminate copies of methods that have a side-affect and non-side affect version: example: moved and move

Just so that it's clear in my mind, this means that methods that act on an object will return a copy of the object instead of mutating it. Is that correct?

> 7. when doing versioning, lets follow semver.org

Just making a note that we (including me) have not always been consistent with this, and our current tag versions don't conform to those standards.

> 8. prefer kwargs for functions/methods

I'm for this, but I could see there being an argument that kwargs can/will make it harder to know what parameters functions can take.

> 9. Let's try to solve the confusion with Vector, Location, Tuple, etc. Users commonly complain that there doesnt seem to be a standard, so its confusing.

There are multiple issues about Vector vs Tuple such as the one [here](https://github.com/CadQuery/cadquery/issues/256), and whenever I've updated code or created new features I've tried to write functions so that Vectors and Tuples were interchangeable. I don't know if that's the intent with this item, or if there is something deeper intended here.

**Reply by dcowden** [reply]
> Just so that it's clear in my mind, this means that methods that act on an object will return a copy of the object instead of mutating it. Is that correct?

that's what i had i mind yes

> There are multiple issues about Vector vs Tuple such as the one https://github.com/CadQuery/cadquery/issues/256, and whenever I've updated code or created new features I've tried to write functions so that Vectors and Tuples were interchangeable. I don't know if that's the intent with this item, or if there is something deeper intended here.

I'd like to make an affirmative design rule about what we will do, and do it everywhere.  I dont have a srong opnion on always vector, alwaqys tuple, or always accept both, but i think users want whichever of three to be done consistently. I'll update this verbiage

**Comment by dcowden** [comment]
Hearing no more comments, closing this as finished
